appserver/web/weld-integration/src/main/java/org/glassfish/weld/ACLSingletonProvider.java

```diff
@@ -86,8 +88,45 @@ public <T> ACLSingleton<T> create(Class<? extends T> expectedType) {
     private static class ACLSingleton<T> implements Singleton<T> {
 
       private final Map<ClassLoader, T> store = new ConcurrentHashMap<>();
-      private final Map<String, T> storeById = new ConcurrentHashMap<>();
+      private final Map<ClassLoaderAndId, T> storeById = new ConcurrentHashMap<>();
       private ClassLoader ccl = Globals.get(ClassLoaderHierarchy.class).getCommonClassLoader();
+      private final Deployment deployment = Globals.getDefaultHabitat().getService(Deployment.class);
+
+        private static class ClassLoaderAndId {
+            private final ClassLoader cl;
+            private final ClassLoader backupClassLoader;
+            private final String id;
+
+            ClassLoaderAndId(String id) {
+                this.id = id;
+                this.cl = this.backupClassLoader = null;
+            }
+
+            ClassLoaderAndId(String id, ClassLoader cl) {
+                this.id = id;
+                this.cl = this.backupClassLoader = cl;
+            }
+
+            ClassLoaderAndId(String id, ClassLoader cl, ClassLoader backupClassLoader) {
+                this.id = id;
+                this.cl = cl;
+                this.backupClassLoader = backupClassLoader;
+            }
+
+            @Override
+            public boolean equals(Object o) {
+                if (!(o instanceof ClassLoaderAndId)) return false;
+                ClassLoaderAndId that = (ClassLoaderAndId) o;
+                return Objects.equals(id, that.id)
+                        && (cl == null || Objects.equals(cl, that.cl)
```

Comment:
Not saying it's important or that it even matters, but this implementation does not seem to be _symmetric_. Also - I don't know how exactly `store.containsKey` below works.

Reply:
You are right, it's not symmetric. This is by design. This class has only a specific use case, and the `equals()` supports this.

Reply:
I did a bit of research around this topic, and it's probably not acceptable to leave it this way.
It does rely on an implementation detail of `ConcurrentHashMap` (or `HashMap` in general) and probably works by coincidence. I'l need to fix this somehow :(

Reply:
Apparently, Weld has a version of `Multimap` that I've used to refactor this into idiomatic code.
Thank you Piotrek